mutation_reader.cc

```diff
@@ -303,6 +305,37 @@ struct mutation_reader_merger::fragment_heap_compare {
     }
 };
 
+void mutation_reader_merger::maybe_add_readers_at_partition_boundary() {
+    // We are either crossing partition boundary or ran out of
+    // readers. If there are halted readers then we are just
+    // waiting for a fast-forward so there is nothing to do.
+    if (_fragment_heap.empty() && _halted_readers.empty()) {
+        if (_reader_heap.empty()) {
+            maybe_add_readers(std::nullopt);
+        } else {
+            maybe_add_readers(_reader_heap.front().fragment.as_partition_start().key());
+        }
+    }
+}
+
+void mutation_reader_merger::retire_reader(reader_and_last_fragment_kind rk) {
```

Comment:
This extraction doesn't seem necessary in the new version, but if you think it makes the code easier to read we can still do it, but in that case it should be in a commit saying it is extracted for readability.

Reply:
OK, I inlined `retire_reader` back. I left `maybe_add_readers_at_partition_boundary` intact, and included extraction of `prepare_one` in the first commit.